### PR TITLE
MSSQL Client Encryption support

### PR DESCRIPTION
**Pull request by tsegismont**
See individual commits for easier review

- [x] encryption level negotiation
- [x] encryption implementation
- [x] manual testing with Docker
- [x] manual testing with Azure database (see [comment](https://github.com/eclipse-vertx/vertx-sql-client/pull/1017#issuecomment-896565891))
- [x] automated testing with Docker
- [x] documentation

### CODE REVIEW

File: vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/codec/TdsSslHandshakeCodec.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    @Override
+    protected Object decode(ChannelHandlerContext ctx, ByteBuf in) throws Exception {
+      ByteBuf byteBuf = (ByteBuf) super.decode(ctx, in);
+      if (byteBuf == null) {
+        return null;
+      }
+
+      int length = in.getUnsignedShort(2);
+
+      return byteBuf.slice(PACKET_HEADER_SIZE, length - PACKET_HEADER_SIZE);
```
Comment (vietj) [inline vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/codec/TdsSslHandshakeCodec.java:52]: as a slice is returned, it is not clear when the original `ByteBuf` will be released as we don't keep a reference on it.

File: vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/codec/TdsSslHandshakeCodec.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    }
+  }
+
+  private static class Encoder extends ChannelOutboundHandlerAdapter {
+
+    private ByteBufAllocator alloc;
+    private CompositeByteBuf accumulator;
```
Comment (vietj) [inline vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/codec/TdsSslHandshakeCodec.java:59]: it does not seem that the accumulator is released on a failure of the pipeline

**Review by vietj — COMMENTED** [review]
a few concerns about potential buffer leaks

**Comment by tsegismont** [comment]
Thanks @vietj , I will address the concerns asap

**Comment by tsegismont** [comment]
### Manual testing with Azure results

Connecting from the outside world works but not from Azure Web App Service.

Did some debugging and it seems this happens because our client doesn't support the ROUTING token of a login response, unlike the JDBC client.

So this PR only partially fix the issue with connecting to Azure SQL servers.

File: vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/codec/TdsSslHandshakeCodec.java (hunk 1 above)
Comment (tsegismont) [reply vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/codec/TdsSslHandshakeCodec.java:52]: The slice instance `release` method is a delegate to the `release` method of the original `ByteBuf`. So when the slice is released (by the `SslHandler`, which comes next in the pipeline) the original `ByteBuf` will be released too.

File: vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/codec/TdsSslHandshakeCodec.java (hunk 2 above)
Comment (tsegismont) [reply vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/codec/TdsSslHandshakeCodec.java:59]: I thought this case was handled by the `handlerRemoved` method. Do you have another case in mind?

Comment (vietj) [reply vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/codec/TdsSslHandshakeCodec.java:59]: right I missed that

**Review by vietj — APPROVED** [review]

**Comment by tsegismont** [comment]
@vietj PTAL, I added a final commit with doc + examples.

**Comment by tsegismont** [comment]
@vietj do you want to review the last commit with documentation or should I go ahead and merge?

**Review by vietj — APPROVED** [review]
LGTM, however I think we should summarize the SSL setup interactions somewhere (javadoc is fine) so anyone debugging can understand how the piece interact together

**Comment by tsegismont** [comment]
good point, I will do that and merge

**Pull request merged**